src/api/wallet/WalletApi.ts

```diff
@@ -293,28 +337,59 @@ export class WalletApiImpl implements WalletApi {
 
   /* ****************      Private Functions      **************** */
 
+  private async _getAsyncWalletInfo(): Promise<WalletInfo> {
+    try {
+      const [[userAddress], networkId] = await Promise.all([this._web3.eth.getAccounts(), this._web3.eth.net.getId()])
+
+      return {
+        userAddress,
+        networkId,
+        isConnected: !!userAddress && !!networkId,
+      }
+    } catch (error) {
+      log('Error asynchrously getting WalletInfo', error)
+      return {
+        userAddress: '',
+        networkId: 0,
+        isConnected: false,
+      }
+    }
+  }
+
   private async _notifyListeners(blockchainUpdate?: BlockchainUpdatePrompt): Promise<void> {
     if (blockchainUpdate) this.blockchainState = blockchainUpdate
 
     await Promise.resolve()
-    const walletInfo: WalletInfo = this.getWalletInfo()
+
+    const walletInfo = await (this.getWalletInfo() || this._getAsyncWalletInfo())
```

Comment:
I don't understand why add this kind of complications. I prefer if the Api interface return a single type, in this case, a `Promise<WalletInfo>`

This logic seems very strange

Reply:
I didn't want to drop the very useful case of sychronous WalletInfo
Most of the time it's synchronous, only with some providers (like TrustWallet) we can't count on synch  props on provider itself

Reply:
Yep, but as soon as 1 provider does that async, we need to make the interface async to avoid confusion.

All parts using this, just need to do a `await` if they need synchronous code. I don't see the benefit in the "dual signature". Maybe I'm not seeing sth...

Reply:
The benefit comes from using it in component render

Reply:
If you will strongly about it, can of course change to always Promise

Reply:
But the issue I see, is that, you cannot use it in render anymore, since you need to handle both cases. So I don't understand the gain

Reply:
```ts
export const useWalletConnection = (): (WalletInfo & { pending: false }) | typeof PendingState => {
  const [walletInfo, setWalletInfo] = useSafeState(() => walletApi.getWalletInfo())

  useEffect((): Command => {
    return walletApi.addOnChangeWalletInfo(setWalletInfo)
  }, [setWalletInfo])

  return useMemo(() => {
    return isPromise(walletInfo) ? PendingState : { ...walletInfo, pending: false }
  }, [walletInfo])
}
```

`isPromise()===false` we return walletInfo, true -- pending

Reply:
You do feel strongly about it, don't you
I'm happy to change

Reply:
Heheh, yes. But may only be cause I don't fully understand the benefits and I'm seeing the problems on an Api that don't return a pure type.

Happy to discuss too, I might be terribly wrong u might convince me